### ISSUE TITLE
Add FileCheck annotations to MIR-opt tests

### ISSUE THREAD

**Issue by cjgillot** [rust-lang/rust#116971]
The MIR-opt testsuite (`tests/mir-opt`) supports checking the MIR output using LLVM FileCheck tool. The purpose of this check is to write in the `.rs` file what we expect to see, similar to the annotations we have for ui tests. See individual commits in https://github.com/rust-lang/rust/pull/116810 for a few examples, and `tests/mir-opt/README.md` for tips in dealing with functions, locals and basic blocks.

Most testcases disable this using a `// skip-filecheck` comment. They should not.
We need to go through all the test cases in the MIR-opt suite to write these annotations.
Exception: the `tests/mir-opt/pre-codegen` does not need these directives, as its purpose is to track the quality of the MIR we emit.

Steps, for each test case:
- figure out what is being tested (that's sometimes not directly obvious, may need git archeology);
- write the FileCheck annotations based on the emitted MIR (see the MIR dumps to help).

In some cases, it may be interesting to strengthen the test cases to better verify what they meant to verify. This can be done by adding a `// unit-test: SomePassName` directive at the top, or changing compiler flags.

The point is to add check directives for what we want to check. Skipping storage statements, basic-block numbers, useless assignments is encouraged.

For volunteers: please post a comment with the files you want to work with. Please do not claim the whole issue, there is plenty to do. Please `r? cjgillot` in the PR so I can keep track of them.

**Comment by rmehri01** [comment]
I'll try working on `tests/mir-opt/inline` :)

**Comment by DavidCoy77** [comment]
Hi! I'd love to get started contributing to Rust, but I'd need a little help getting started since this would be my first ever GitHub contribution. You've included the tags Easy, and Help-Wanted, so since there are plenty of files to work on besides the one @rmehri01 claimed, if you could help me out I'm eager to get to work.

**Comment by rmehri01** [comment]
Hey there, I can offer a bit of insight but maybe someone else can add more details :)

For getting started check out [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) if you haven't already, the rustc dev guide is really comprehensive and helpful! 

For this issue specifically, I basically went to each file in the above directory and went through the GitHub history (related issues and PRs, especially the initial PR that created the test file) to try to determine what is trying to be tested. The diff files for each test are also really helpful for this since they show what changes the test actually made. Along with the file name and comments, you can usually get an idea of what to test. Then you can use the [LLVM FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html) syntax to check the expected result.

Feel free to reach out if you're stuck, I'll try my best to help.

**Comment by DavidCoy77** [comment]
I'm still a beginner so I definitely appreciate the willingness to help! Let me know if I have a good handle on the issue: the tests/mir-opt/ directory contains code for various tests you can run from the command line to test MIR optimizations. After some minor "git-archaeology" it seems the pattern  is the `// skip-filecheck ` needs to be added to the top of the .rs files, and the filecheck annotation code needs to be added to the corresponding .diff file. 

You mention running the FileCheck test, and getting an idea of what to test, but how would I go about that? Additionally, I don't quite understand the code in the .diff files. Again I'm only a beginner so I appreciate you time and effort!

**Comment by rmehri01** [comment]
Not quite, the `// skip-filecheck` is actually there right now to disable running FileCheck:

> Most testcases disable this using a `// skip-filecheck` comment. They should not.

And we want to add the FileCheck annotations to the source `.rs` code, the diffs are just helpful for seeing what changes due to the MIR optimization.

You can check out the commits in the PR mentioned in the issue or in my PR for some examples. You want to remove that `// skip-filecheck` line so that FileCheck does run on the file. As for running FileCheck, this is already done in the previous PR mentioned in this issue, so running the tests with `./x test tests/mir-opt --stage 1` for example will run FileCheck automatically. For the code in the diff files, https://rustc-dev-guide.rust-lang.org/mir/index.html (the linked blog post is quite helpful) and https://rustc-dev-guide.rust-lang.org/mir/optimizations.html may help.

**Comment by tjausm** [comment]
I'll start on the `tests/mir-opt/const_prop/` folder

**Comment by sasurau4** [comment]
I'll start on the `tests/mir-opt/copy-prop/` folder

**Comment by Tyrubias** [comment]
I'll work on the `tests/mir-opt/dest-prop/` folder.

**Comment by sfzhu93** [comment]
I've started on directory `tests/mir-opt/dataflow-const-prop`.

**Comment by sfzhu93** [comment]
Just some experience to share: we should test against 32 bit and 64 bit platforms, as well as platforms with different unwinding. `wasm32-unknown-unknown` is helpful because it differs from `x86_64-unknown-linux-gnu` in both bits and unwinding.

**Comment by JarlEvanson** [comment]
I'll work on the `tests/mir-opt/sroa` folder.

**Comment by CastilloDel** [comment]
I'll work on the `tests/mir-opt/dest-prop` folder

**Comment by Shunpoco** [comment]
I'll work on `tests/mir-opt/issues` and `tests/mir-opt/copy-prop`

**Comment by laurenmchin** [comment]
I'd like to work on adding FileCheck annotations for `tests/mir-opt/dead-store-elimination` and `tests/mir-opt/instsimplify`. I'll analyze the emitted MIR, add the relevant check directives, and strengthen the test cases as needed. Thanks!

**Comment by moatom** [comment]
I'll work on the `tests/mir-opt/nll/` folder. (This is the last one, isn't it?")

**Comment by laurenmchin** [comment]
> I'd like to work on adding FileCheck annotations for `tests/mir-opt/dead-store-elimination` and `tests/mir-opt/instsimplify`. I'll analyze the emitted MIR, add the relevant check directives, and strengthen the test cases as needed. Thanks!

Hello! Apologies for the delay-I’ve carved out time to work on these during the last week of March and will update soon. Thanks for your patience!

**Comment by joseph-isaacs** [comment]
PRs for 
- `matches_u8.rs` pr: https://github.com/rust-lang/rust/pull/162068
- `uninhabited_fallthrough_elimination.rs` pr: https://github.com/rust-lang/rust/pull/162096
- `const_prop/large_array_index.rs.rs` pr: https://github.com/rust-lang/rust/pull/162095
